gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -0,0 +1,216 @@
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.batching.v2;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.BetaApi;
+import com.google.api.core.InternalApi;
+import com.google.api.core.SettableApiFuture;
+import com.google.api.gax.rpc.UnaryCallable;
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Queues up the elements until {@link #flush()} is called; once batching is over, returned future
+ * resolves.
+ *
+ * <p>This class is not thread-safe, and expects to be used from a single thread.
+ */
+@BetaApi("The surface for batching is not stable yet and may change in the future.")
+@InternalApi
+public class BatcherImpl<ElementT, ElementResultT, RequestT, ResponseT>
+    implements Batcher<ElementT, ElementResultT> {
+
+  private final BatchingDescriptor<ElementT, ElementResultT, RequestT, ResponseT>
+      batchingDescriptor;
+  private final UnaryCallable<RequestT, ResponseT> callable;
+  private final RequestT prototype;
+  private Batch<ElementT, ElementResultT, RequestT> currentOpenBatch;
+
+  private final AtomicInteger numOfOutstandingBatches = new AtomicInteger(0);
+  private final Object flushLock = new Object();
+  private boolean isClosed = false;
+
+  private BatcherImpl(Builder<ElementT, ElementResultT, RequestT, ResponseT> builder) {
+    this.prototype = checkNotNull(builder.prototype, "RequestPrototype cannot be null.");
+    this.callable = checkNotNull(builder.unaryCallable, "UnaryCallable cannot be null.");
+    this.batchingDescriptor =
+        checkNotNull(builder.batchingDescriptor, "BatchingDescriptor cannot be null.");
+  }
+
+  /** Builder for a BatcherImpl. */
+  public static class Builder<ElementT, ElementResultT, RequestT, ResponseT> {
```

Comment:
can this be `@Autovalue`? I'm not a big fan of `@Autovalue` but we already use it extensively in gax-java, so lets reuse it here if possible to reduce boilerplate code typing.

Reply:
I don't BatcherImpl is a value object

Reply:
I don't think it has to be a value object, it can be anything, which can be constructed with a builder. 
Now when I look at it, do we even need a builder here? I'm just trying to figure out if we can reduce boilerplate code here (either by generating builder with the autovalue annotation or not having it at all).

Reply:
I dont think a builder is necessary here. Given the choice between AutoValue and removing the builder I would prefer removing the builder

Reply:
I have removed the Builder from the current change and annotated this class with `@AutoValue`.

Just a side note, We may need to come back on this when introducing flow controllers. As after implementing triggers, we would be expecting 6 arguments(i.e. existing ones + BatchingSettings, FlowController, and an Executor).

Reply:
TLDR version: please remove AutoValue for the time being.


Longer version:
I've given this more thought, and I really think that AutoValue is really the  wrong tool for this. In the [best practices](https://github.com/google/auto/blob/master/value/userguide/practices.md#equals-means-interchangeable) it states

> Don't use AutoValue to implement value semantics unless you really want value semantics. In particular, you should never care about the difference between two equal instances.

This is definitely not the case for BatcherImpl since a Batcher has mutable state and we do care which instance we call flush on. Also, using AutoValue here implies that equality will use value semantics and be determined by the prototype, callable & batchingDescriptor:

```java
BatcherImpl.Builder builder = ...;

Batcher b1 = builder.build();
Batcher b2 = builder.build();
b2.add(...)

b1.equals(b2) == true; 
```
Which makes no sense. I think the only meaningful equality here is referential and goes against the best practices of AutoValue.

Since you already removed the Builder until a later PR, please remove the use of AutoValue here since it buys you very little. Once that PR comes up I'll chat with vam about the best path forward.